File: Data Structures & Algorithms/islands-and-treasure/submission-2.cpp
class Solution {
    
public:
    void islandsAndTreasure(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        
        queue<pair<int,int>> q;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 0) {
                    q.push({i,j});
                }
            }
        }

        if (q.size() == 0) {
            return;
        }
        int dx[4]  = {1,-1,0,0}; 
        int dy[4]  = {0,0,1,-1}; 
        auto isValid = [&](int i, int j) {
            return (i >= 0 and i < n and j >= 0 and j < m);
        };
        while(q.size() > 0) {
            int row = q.front().first;
            int col = q.front().second;
            q.pop();
            
            for (int i = 0; i < 4; i++) {
                int nr = row + dx[i];
                int nc = col + dy[i];
                if (!isValid(nr,nc) or grid[nr][nc] != INT_MAX) {
                    continue;
                }
                q.push({nr,nc});
                grid[nr][nc] = grid[row][col] + 1;
            }
        }

    }
};

File: Data Structures & Algorithms/rotting-fruit/submission-2.cs
public class Solution {
    bool IsValid(int i, int j , int n , int m) {
        return (i >= 0 && i < n && j >= 0 && j < m);
    }
    public int OrangesRotting(int[][] grid) {
        int n = grid.Length;
        int m = grid[0].Length;
        Queue<int[]> q = new Queue<int[]>();
        int fresh = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 2) {
                    q.Enqueue(new int[]{i,j});
                }
                if (grid[i][j] == 1) fresh++;
            }
        }
     

        int[]dx = {1,-1,0,0};
        int[]dy = {0,0,1,-1};
        int ans = 0;

        while(fresh > 0 && q.Count > 0) {
            int len = q.Count;
            for (int j = 0; j < len; j++) {
                var cur = q.Dequeue();
                int row = cur[0], col = cur[1];

                for (int i = 0; i < 4; i++) {
                    int nr = row + dx[i];
                    int nc = col + dy[i];
                    if (IsValid(nr,nc,n,m) && grid[nr][nc] == 1) {
                        fresh--;
                        grid[nr][nc] = 2;
                        q.Enqueue(new int[]{nr,nc});
                    }
                }
            }
            ans++;
            
        }
        return (fresh == 0 ? ans : -1);
    }
}
